src/ir/possible-contents.cpp

```diff
@@ -0,0 +1,1607 @@
+    //       reading from the immutable global"
+    assert(refContents.isMany() || refContents.isGlobal());
+
+    // We create a special location for the canonical cone of this type, to
+    // avoid bloating the graph, see comment on Flower::canonicalConeReads().
+    // TODO: A cone with no subtypes needs no canonical location, just
+    //       add one direct link here.
+    auto& coneReadIndex = canonicalConeReads[std::pair<HeapType, Index>(
+      declaredHeapType, fieldIndex)];
+    if (coneReadIndex == 0) {
+      // 0 is an impossible index for a LocationIndex (as there must be
+      // something at index 0 already - the ExpressionLocation of this very
+      // expression, in particular), so we can use that as an indicator that we
+      // have never allocated one yet, and do so now.
+      coneReadIndex = makeSpecialLocation();
+      for (auto type : subTypes->getAllSubTypes(declaredHeapType)) {
+        connectDuringFlow(DataLocation{type, fieldIndex},
+                          SpecialLocation{coneReadIndex});
+      }
+
+      // TODO: if the old contents here were an exact type then we have an old
+      //       link here that we could remove as it is redundant (the cone
+      //       contains the exact type among the others). But removing links
+      //       is not efficient, so maybe not worth it.
+    }
+
+    // Link to the canonical location.
+    connectDuringFlow(SpecialLocation{coneReadIndex},
+                      ExpressionLocation{read, 0});
+  }
+}
+
+void Flower::writeToData(Expression* ref, Expression* value, Index fieldIndex) {
+#if defined(POSSIBLE_CONTENTS_DEBUG) && POSSIBLE_CONTENTS_DEBUG >= 2
+  std::cout << "    add special writes\n";
+#endif
+
+  // We could set up links here as we do for reads, but as we get to this code
+  // in any case, we can just flow the values forward directly. This avoids
+  // increasing the size of the graph.
+  //
+  // Figure out what to send in a simple way that does not even check whether
+  // the ref or the value was just updated: simply figure out the values being
+  // written in the current state (which is after the current update) and
+  // forward them.
+  auto refContents = getContents(getIndex(ExpressionLocation{ref, 0}));
+  auto valueContents = getContents(getIndex(ExpressionLocation{value, 0}));
+  if (refContents.isNone() || refContents.isNull()) {
+    return;
+  }
+  if (refContents.hasExactType()) {
+    // Update the one possible type here.
+    auto heapLoc =
+      DataLocation{refContents.getType().getHeapType(), fieldIndex};
+    updateContents(heapLoc, valueContents);
+  } else {
+    assert(refContents.isMany() || refContents.isGlobal());
+
+    // Update all possible subtypes here.
+    auto type = ref->type.getHeapType();
+    for (auto subType : subTypes->getAllSubTypes(type)) {
+      auto heapLoc = DataLocation{subType, fieldIndex};
+      updateContents(heapLoc, valueContents);
+    }
+  }
```

Comment:
I am still confused.

1. I guess you mean `DataLocation` by `StructLocation`, right? (There's no `StructLocation`)

2. Why do we use `SpecialLocation`? I think it makes sense if `SpecialLocation`s are uniqued, but it looks they are created every time we process `struct.get` / `array.get` (per subtype): https://github.com/WebAssembly/binaryen/pull/4685/files#diff-4d9d6d08d863caa22d1fa86a99f2b7d37d1013d36c98e177aca6edd6eaff54e1R1489
Then what makes this more efficient? For example, a `struct.get`'s field 0 can have 3 subtypes. Let's call them A, B, and C. And let's call the read expression's location E. And we created a special location S. If we do the optimization, we will create connections of A->S, B->S, and C->S, and then S->E. If we don't do the optimization, we will create A->E, B->E, and C->E. And for another `struct.get`, we will create another special location S2, and then this will repeat. What's the benefit? And we will even do this multiple times for the same `struct.get`, every time this `struct.get` is pushed to the `workQueue`.

3. About why we don't use `connectDuringFlow`, `connectDuringFlow` adds new links that will update targets later. It's OK not to do that in `writeToData`?

4. Why do we do this (creating `SpecialLocation` and stuff) later? Can't we do this when we first run `InfoCollector`? It looks simpler, and in the current way it looks we end up doing the same structure building (creating special locations and links and such) for expressions like `struct.get`/`struct.set` every time these exprs are updated.

Reply:
2. Ah, right, I missed that we query `canonicalConeReads` map before creating a new `SpecialLocation`. Thanks.

3. `readFromData` uses `connectDuringFlow` but `writeToData` does not.

---
Also I still don't understand your earlier explanation on why we use `SpecialLocation` only in `readFromData` and not in `writeToData`. As you pointed out, if # of subtypes is `n` and `readFromData` is called from `m` `struct.get`s, if we use `SpecialLocation`s, the # of edges is `O(n+m)`, but if we use `SpecialLocation`s, the # of edges is `O(n*m)`. (Assume we only think about a single field index, or an `array.get`.)

But doesn't this apply in the same way to `writeToData`? If # of subtypes is `n` and `writeToData` is called from `m` `struct.set`s, in the current impl, the # of edges is `O(n*m)`, and it can be reduced to `O(n+m)` if we use `SpecialLocation`s there too.

Reply:
Maybe a way to get an intuition for why read is not symmetric with write here is this: When we do a write, we are sending data forward. A read not only reads stuff right now, but also must receive data in the future. So it needs to modify the graph, which a write doesn't.

In more detail, in a write we know exactly where to send it to: if it's a precise type in the ref of the write, then exactly one `DataLocation` (for that exact type, and its field). If it's not precise, then all N subtypes. We **must** do that work, since our data representation looks like `DataLocation{type, fieldIndex}`. So we have no choice but to do N operations here (at least with this representation). But, this is not so bad actually :smile: It is N operations, but at least **the graph size** is not N, because we can just do a simple loop on the subtypes, and we're done. We don't need to add anything to the graph.

Reads are much harder, because we are not just passing data forward. We could do a loop over subtypes here (if we need subtypes), and read from those `DataLocations` right now, but that is not enough: we also need future updates to those `DataLocations` to reach us. For that, we need to add connections to the graph. This is where things can start to take lots of memory: we want to avoid adding N connections for every `struct.get`, and we use canonical cone reads to avoid that: one canonical cone read has N connections while the M `struct.get`s have just one (to the canonical one).

Reply:
I asked this in the previous comment but I'm not sure if this was answered... `readFromData` uses `connectDuringFlow`, but `writeToData` doesn't. Why?

> Maybe a way to get an intuition for why read is not symmetric with write here is this: When we do a write, we are sending data forward. A read not only reads stuff right now, but also must receive data in the future. So it needs to modify the graph, which a write doesn't.

I'm not sure what you mean by "sending data forward" vs. "receive data in the future". Does "forward" here man the past, because you contrasted it with "future"?

And why does this "receive from future" matter only for `struct.set`/`array.set`? Isn't this the same for all expressions? Like, if any expression's child gets a new `PossibleContents` at some point in the future, the expression's `PossibleContents` should be updated... No?

I'm having a hard time understanding why they are not symmetric, and the concepts of forward vs. future.

> But, this is not so bad actually 😄 It is N operations, but at least **the graph size** is not N, because we can just do a simple loop on the subtypes, and we're done. We don't need to add anything to the graph.

I'm not sure what "the graph size" here means.

I guess this is taking too long, sorry, my understanding is bad. 😢 Maybe it's better to take remaining things offline..

Reply:
We talked about this offline. Thanks @kripken for the patient explanation! I thought about this a little more after the meeting and now I think I understand things better! I'll start with summarizing what I understood, to make sure what I understood is correct. (Let me know if not.)

- The reason we don't pre-create edges with `DataLocation`s for all possible subtypes for `struct.get`/`struct.set` (I'll omit `array.get` and `array.set`, but think of them as included too) is it can create unnecessary many edges for each of possible subtypes in places in which we possibly only end up with only an `ExactType`.
- The reason `writeToData` doesn't use `connectDuringFlow` is we don't want to increase the size of the graph, more precisely the # of edges in the graph. We transfer contents from `value` -> `ref`'s heap type's `DataLocation`, but we don't need to _materialize_ the connections, meaning adding them to `links`, to do this flowing, because we can do that manually each time a `struct.set` is visited in `updateContents`.
- The reason `readFromData` is different from `writeToData` is, `struct.get`'s destination expression needs to be updated each time its heap type's (+ possibly its subtypes') `DataLocation`s are updated. (This is different from other expressions because other expressions generally need to be updated when one of their children is updated.) So we create dynamic connections from the heap type's `DataLocation` -> `struct.get`'s `ExpressionLocation` and, unlike `writeFromData`, also _materialize_ them (= add them to `links`), to _subscribe_ to future updates from the heap type's (+ possibly its subtypes') `DataLocation`s. (These `DataLocation` updates will be done when updating `struct.set`s in the future.) But this can create `O(n*m)` edges; so we introduce `SpecialLocation` that represents cone types, with which we only need to create `O(n+m)` edges.